crates/lang_srv/src/analysis/completion.rs

```diff
@@ -0,0 +1,443 @@
+            }
+            roc_types::subs::FlatType::Tuple(elems, ext) => {
+                let elems = elems.unsorted_iterator(subs, ext);
+
+                match elems {
+                    Ok(elem) => elem.map(|(num, var)| (num.to_string(), var)).collect(),
+                    Err(err) => {
+                        warn!("Error getting tuple elems for completion: {:?}", err);
+                        vec![]
+                    }
+                }
+            }
+
+            _ => {
+                warn!(
+                    "Trying to get field completion for a type that is not a record: {:?}",
+                    typ
+                );
+                vec![]
+            }
+        },
+        roc_types::subs::Content::Error => {
+            //This is caused by typechecking our partially typed variable name causing the typechecking to be confused as the type of the parent variable
+            //TODO! ideally i could recover using some previous typecheck result that isn't broken
+            warn!("Variable type of record was of type 'error', cannot access field",);
+            vec![]
+        }
+        _ => {
+            warn!(
+                "Variable before field was unsupported type: {:?}",
+                subs.dbg(var)
+            );
+            vec![]
+        }
+    }
+}
+
+struct FieldCompletion {
+    var: String,
+    field: String,
+    middle_fields: Vec<String>,
+}
+///Splits a completion prefix for a field into its components
+///E.g. a.b.c.d->{var:"a",middle_fields:["b","c"],field:"d"}
+fn get_field_completion_parts(symbol_prefix: &str) -> Option<FieldCompletion> {
+    let mut parts = symbol_prefix.split('.').collect::<Vec<_>>();
+    let field = parts.pop().unwrap_or("").to_string();
+    let var = parts.remove(0);
+    //Now that we have the head and tail removed  this is all the intermediate fields
+    let middle_fields = parts.into_iter().map(ToString::to_string).collect();
+
+    Some(FieldCompletion {
+        var: var.to_string(),
+        field,
+        middle_fields,
+    })
+}
+pub fn field_completion(
+    position: Position,
+    symbol_prefix: String,
+    declarations: &Declarations,
+    interns: &Interns,
+    subs: &mut Subs,
+    module_id: &ModuleId,
+) -> Option<Vec<CompletionItem>> {
+    let FieldCompletion {
+        var,
+        field,
+        middle_fields,
+    } = get_field_completion_parts(&symbol_prefix)?;
+
+    debug!(
+        "getting record field completions: variable: {:?} field: {:?} middle: {:?} ",
+        var, field, middle_fields
+    );
+    //TODO: this is kind of just a hack. We are getting all the completions and seeing if any match the part before the dot as a way to get the Variable type of the variable before the dot. I imagine there are much faster ways to do this
+    let completion = get_completions(position, declarations, var.to_string(), interns)
+        .into_iter()
+        .map(|a| (a.0.as_str(interns).to_string(), a.1))
+        .next()?;
```

Comment:
This is fine, this is not a hack. Let's get rid of this comment. We don't need a specialized data structure for this.